add functions & unit tests for converting gDR MAE to PSet

jjjermiah opened the pull request:
# Description
## What changed?
 Functions to convert gDR-generated MAE to a PharmacoSet.


## Why was it changed?
Request by Marc & Benjamin.  

# Checklist for sustainable code base
- [ X ] I added tests for any code changed/added
- [ X ] I added documentation for any code changed/added
- [ X ] I made sure naming of any new functions is self-explanatory and consistent

# Logistic checklist
- [ ] Package version bumped
- [ ] Changelog updated

# Screenshots (optional)

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+                                pset_name) {
+    checkmate::assert_class(mae, "MultiAssayExperiment")
+    checkmate::assert_string(pset_name)
+    # checkmate::assert_named(mae, "assays")
+    # checkmate::assert_named(mae, "colData")
+    # checkmate::assert_named(mae, "rowData")
+    # checkmate::assert_named(mae, "metadata")
```
```suggestion
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+#' @importFrom PharmacoGx PharmacoSet2
+#' 
+#' @examples
+#' # Convert a MultiAssayExperiment object to a PharmacoSet object
+#' pset <- convert_MAE_to_PSet(mae, "my_pset")
```
```suggestion
convert_MAE_to_PSet(mae, "my_pset")
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+#' Convert MultiAssayExperiment to TreatmentResponseExperiment
+#'
+#' This function converts a MultiAssayExperiment generated by gDR into a TreatmentResponseExperiment for use in the PharmacoGx package. The resulting PharmacoSet can be used for pharmacogenomic analysis of drug response.
```
```suggestion
#' This function converts a MultiAssayExperiment generated by gDR into a TreatmentResponseExperiment for use in the 
#' PharmacoGx package. The resulting PharmacoSet can be used for pharmacogenomic analysis of drug response.
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    assay_names <- unique(unlist(gDRutils::MAEpply(mae, function(x) assayNames(x))))
+
+    # Create a list of data.tables, one for each assay name, with the data from all Summarized Experiments in the MAE. 
+    assay_list <- lapply(assay_names, function(x){
```
```suggestion
    assay_list <- lapply(assay_names, function(x) {
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+        # Get list of datatables with all assays with name=x converted to a data.table
+        se_dt_list <- lapply(se_list, FUN = gDRutils::convert_se_assay_to_dt, x)
+
+        # Combine all the data.tables into one data.table, add the name of the Summarized Experiment as a "data_type" column
+        # so that users can identify which Summarized Experiment the data came from
```
```suggestion
        # Combine all the data.tables into one data.table, add the name of the Summarized Experiment
        # as a "data_type" column so that users can identify which Summarized Experiment the data came from
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    assay_dt <- data.table::rbindlist(assay_list, fill = TRUE, idcol = "assay_name")
+
+    # Get the unique rownames from all experiments combined
+    rownames <- unique(unlist(gDRutils::MAEpply(mae, FUN=rownames)))
```
```suggestion
    rownames <- unique(unlist(gDRutils::MAEpply(mae, FUN = rownames)))
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+
+
+    # Get the unique colnames from all experiments combined
+    colnames <- unique(unlist(gDRutils::MAEpply(mae, FUN=colnames)))
```
```suggestion
    colnames <- unique(unlist(gDRutils::MAEpply(mae, FUN = colnames)))
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    # As of this version, curation slot is left empty. 
+    PharmacoGx::PharmacoSet2(
+        name = pset_name, 
+        treatment = as.data.frame(rowdata_), 
+        sample = as.data.frame(coldata_),
```
```suggestion
        treatment = as.data.frame(rowdata_),  # nolint
        sample = as.data.frame(coldata_), # nolint
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+#' @importFrom BumpyMatrix splitAsBumpyMatrix
+#' @importFrom S4Vectors DataFrame
+#'
+.extract_or_create_assay <- function(SE, assay_name){
```
```suggestion
.extract_or_create_assay <- function(SE, 
                                                                assay_name) {
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    } else {
+        SummarizedExperiment::assay(SE, assay_name) <- 
+            BumpyMatrix::splitAsBumpyMatrix(
+                S4Vectors::DataFrame(x = rep(NA,prod(dim(SE)))), 
```
```suggestion
                S4Vectors::DataFrame(x = rep(NA, prod(dim(SE)))), 
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    # rename the rownames column to "sampleid" (required for PharmacoSet Object)
+    data.table::setnames(coldata_, "rn", "sampleid")
+
+    # Create a list of rowname and column name identifiers for each assay name (required for TreatmentResponseExperiment object)
```
```suggestion
    # Create a list of rowname and column name identifiers for each assay name (required for 
    # TreatmentResponseExperiment object)
```

DominikZabinski commented on tests/testthat/test-gDRmae_to_pset.R:
```diff
@@ -0,0 +1,37 @@
+  
+  pset_mae2Exp <- convert_MAE_to_PSet(maeTwoExperiments, "my2exp")
+  expect_equal(class(pset_mae2Exp@treatmentResponse)[1], "TreatmentResponseExperiment")
+  expect_equal(sort(rownames(pset_mae2Exp@treatmentResponse)), sort(unlist(lapply(experiments(maeTwoExperiments), rownames), use.names = FALSE)))
```
```suggestion
  expect_equal(sort(rownames(pset_mae2Exp@treatmentResponse)), 
               sort(unlist(lapply(experiments(maeTwoExperiments), rownames), use.names = FALSE)))
```

DominikZabinski commented on tests/testthat/test-gDRmae_to_pset.R:
```diff
@@ -0,0 +1,37 @@
+  expect_equal(class(pset_mae2Exp@treatmentResponse)[1], "TreatmentResponseExperiment")
+  expect_equal(sort(rownames(pset_mae2Exp@treatmentResponse)), sort(unlist(lapply(experiments(maeTwoExperiments), rownames), use.names = FALSE)))
+  expect_equal(pset_mae2Exp@sample$sampleid, cnames)
+  expect_equal(pset_mae2Exp@treatment$treatmentid, c(rownames( assay(maeTwoExperiments,'single-agent')),rownames(assay(maeTwoExperiments,'matrix'))))
```
```suggestion
  expect_equal(pset_mae2Exp@treatment$treatmentid, 
               c(rownames(assay(maeTwoExperiments, "single-agent")), rownames(assay(maeTwoExperiments, "matrix"))))
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    # Get the unique rownames from all experiments combined
+    rownames <- unique(unlist(gDRutils::MAEpply(mae, FUN=rownames)))
+    # convert the rowData of each Summarized Experiment to a data.table and combine them into one data.table
+    rowdata_ <- data.table::rbindlist(gDRutils::MAEpply(mae, FUN=function(x) data.table::as.data.table(SummarizedExperiment::rowData(x), keep.rownames=TRUE)), fill=TRUE)
```
```suggestion
    rowdata_ <- data.table::rbindlist(
      l = gDRutils::MAEpply(
        mae = mae, 
        FUN = function(x) data.table::as.data.table(SummarizedExperiment::rowData(x), keep.rownames = TRUE)), 
      fill = TRUE)
```

DominikZabinski commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,152 @@
+    # Get the unique colnames from all experiments combined
+    colnames <- unique(unlist(gDRutils::MAEpply(mae, FUN=colnames)))
+    # convert the colData of each Summarized Experiment to a data.table and combine them into one data.table
+    coldata_ <- unique(data.table::rbindlist(gDRutils::MAEpply(mae, FUN=function(x) data.table::as.data.table(SummarizedExperiment::colData(x), keep.rownames=TRUE)), fill=TRUE))
```
```suggestion
    coldata_ <- unique(data.table::rbindlist(
      l = gDRutils::MAEpply(
        mae = mae, 
        FUN = function(x) data.table::as.data.table(SummarizedExperiment::colData(x), keep.rownames = TRUE)), 
      fill = TRUE))
```

DominikZabinski commented on tests/testthat/test-gDRmae_to_pset.R:
```diff
@@ -0,0 +1,37 @@
+
+  pset <- convert_MAE_to_PSet(mae, "my_pset")
+  expect_equal(class(pset@treatmentResponse)[1], "TreatmentResponseExperiment")
+  expect_equal(sort(rownames(pset@treatmentResponse)), sort(unlist(lapply(experiments(mae), rownames), use.names = FALSE)))
```
```suggestion
  expect_equal(sort(rownames(pset@treatmentResponse)), 
               sort(unlist(lapply(MultiAssayExperiment::experiments(mae), rownames), use.names = FALSE)))
```

gladkia commented:
Hi @jjjermiah,

It seems there are some warnings thrown, mostly like:
```
Warning ('test-gDRmae_to_pset.R:15'): convert_MAE_to_PSet works as expected
Arguments in '...' ignored
Backtrace:
  1. gDRimport::convert_MAE_to_PSet(mae, "my_pset")
       at test-gDRmae_to_pset.R:15:2
  3. gDRutils::MAEpply(...)
  4. base::lapply(experiments, FUN, ...)
  5. gDRimport (local) FUN(X[[i]], ...)
  7. data.table:::as.data.table.default(...)
 10. S4Vectors::as.data.frame.Vector(x, ...)
 12. S4Vectors::as.data.frame(...)
 13. S4Vectors (local) .local(x, row.names, optional, ...)
Warning ('test-gDRmae_to_pset.R:15'): convert_MAE_to_PSet works as expected
Arguments in '...' ignored
```

You can debug/fix them on the Docker image. Here is the installation: https://github.com/gdrplatform/gDR#installation procedure. 
Once you have the container running you can work with your gDRimport branch and test it with:
 - testthat::test_local()
 - or with `gDRstyle::checkPackage()`

bczech commented on R/MAE_to_PSet.R:
```diff
@@ -0,0 +1,163 @@
+#' 
+#' @return A PharmacoSet object.
+#' 
+#' @importFrom checkmate assert_class assert_string
```
instead of using `importFrom` please use `package::function` namespacing in the code

jjjermiah replied:
Thanks Bartek, I have made those changes.

jjjermiah commented:
@gladkia I believe another team member fixed those warnings. I have also added a vignette for this conversion. Please check it out and let me know.

gladkia commented:
> @gladkia I believe another team member fixed those warnings. I have also added a vignette for this conversion. Please check it out and let me know.



> @gladkia I believe another team member fixed those warnings. I have also added a vignette for this conversion. Please check it out and let me know.

@jjjermiah - great job! We've added minor corrections to your PR. The builds passed. @bczech tested new functions on real data. We will merge soon.

jjjermiah commented:
> > @gladkia I believe another team member fixed those warnings. I have also added a vignette for this conversion. Please check it out and let me know.
> 
> @jjjermiah - great job! We've added minor corrections to your PR. The builds passed. @bczech tested new functions on real data. We will merge soon.

@gladkia sounds great thank you for your help with this!
